### DISCUSSION TITLE
[Question] Hook/intercep before deserializing json document

### DISCUSSION THREAD

**Discussion by sszakal** [JasperFx/marten#2221] [General] [answered]
Hi all,

Is there a way in MartenDB to intercept a JSON document before it gets converted to CLR object?  (I'm currently on Marten V3).

The reason I'm interested in this type of hook/callback is to apply on-the-fly migrations/transformations to the JSON. I know marten supports JSON transformations using the patch API or using javascript transformations, but these are executed on the database side and it doesn't scale very well and, somehow, limited. Running transformations on the .NET side would allow much more powerful and complex transformations to be performed. 

Please let me know what options do I have?

Thanks,
Stef

**Comment by mysticmind** [comment]
https://martendb.io/diagnostics.html#listening-for-document-store-events - Check `DocumentLoaded` and see if it works for your use case.

https://martendb.io/v3/documentation/documents/diagnostics - v3 docs pertaining to `IDocumentSessionListener`

Edit: I don't think you can modify the loaded docs. This will need to be an enhancement to deal with.

**Comment by sszakal** [comment]
Thanks @mysticmind . 
Yes, IDocumentSessionListener allows to intercept the document before load,  but not before it gets deserialized. 

Are there any plans for adding such hook in IDocumentSessionListener  ?
If I would add such feature myself and open a pull request, where would be the best place to intercept the json stream ? 
I just had a quick look at the source code and I can see few possible places:IdentityMap or , and probably better, in DocumentStorage (Resolve, Fetch  methods) 

Thanks,
Stef

**Comment by mysticmind** [comment]
@sszakal we are happy to take a PR. Note that 3.x is in maintenance mode so we are not adding any new features/enhancements. This feature will need to be in the latest 5.x.

@jeremydmiller Could you please provide some pointers for Stefan to take look at this?

**Reply by jeremydmiller** [reply]
You *could* use a custom ISerializer (the Marten abstraction) that decorates the inner serializer. That's the best that I've got here. 

You also have other options with the Plv8 to transform existing JSON data with javascript functions at one time.

**Reply by sszakal** [reply]
Thank you @jeremydmiller  and @mysticmind .

I initially thought about intercepting the json using a wrapper around the json serializer, but feels a bit of a hack. However, this seems to be the easiest way as I'm currently on v3 and I would have to move my project to latest marten if I would try to add such feature in Marten. 

Thanks again for your help

**Reply by jeremydmiller** [reply]
@oskardudycz is working on event upcasters now that will be in 5.9 pretty soon.

**Reply by oskardudycz** [reply]
The PR https://github.com/JasperFx/marten/pull/2314 is merged already, @sszakal, have a look at the upcoming feature docs: https://630b6e791b8fa54aba654d0a--lucid-northcutt-013a56.netlify.app/events/versioning.html. It'll allow performing raw JSON transformations both for Newtonsoft Json.NET and System.Text.Json.

**Reply by sszakal** [reply]
Thanks @oskardudycz I’m looking forward for this feature !

**Reply by oskardudycz** [reply]
@sszakal, I think that 5.9 should be released pretty soon, we'll ping you when it's out :)

**Reply by oskardudycz** [reply]
@sszakal 5.9.0 is out, see the docs https://martendb.io/events/versioning.html :)

**Comment by oskardudycz** [comment, marked as answer]
In version 5.9.0, we added Upcaster functionality that allows transforming events' JSON payloads on the fly. Read more in https://martendb.io/events/versioning.html#upcasting-advanced-payload-transformations 🙂
